### PR TITLE
add kernel for cumsum and cumprod functions

### PR DESCRIPTION
**Pull request by KsanaKozlova**

### CODE REVIEW

**Review by densmirn — APPROVED** [review]

File: dpnp/backend/kernels/dpnp_krnl_mathematical.cpp [hunk 1]
```diff
@@ -103,6 +103,44 @@ void dpnp_cross_c(void* array1_in, void* array2_in, void* result1, size_t size)
     return;
 }
 
+template <typename _KernelNameSpecialization1, typename _KernelNameSpecialization2>
+class dpnp_cumprod_c_kernel;
+
+template <typename _DataType_input, typename _DataType_output>
+void dpnp_cumprod_c(void* array1_in, void* result1, size_t size)
+{
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_mathematical.cpp:111]:
```suggestion
{
    if (!size)
    {
        return;
    }
```

File: dpnp/backend/kernels/dpnp_krnl_mathematical.cpp [hunk 2]
```diff
@@ -103,6 +103,44 @@ void dpnp_cross_c(void* array1_in, void* array2_in, void* result1, size_t size)
...
+
+    return;
+}
+
+template <typename _KernelNameSpecialization1, typename _KernelNameSpecialization2>
+class dpnp_cumsum_c_kernel;
+
+template <typename _DataType_input, typename _DataType_output>
+void dpnp_cumsum_c(void* array1_in, void* result1, size_t size)
+{
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_mathematical.cpp:130]:
```suggestion
{
    if (!size)
    {
        return;
    }
```

File: dpnp/dpnp_algo/dpnp_algo_mathematical.pyx [hunk 3]
```diff
@@ -344,6 +320,24 @@ cpdef dparray dpnp_multiply(dparray x1, x2):
...
+    for i in range(x1.size):
+        if dpnp.isnan(x1[i]):
+            x1._setitem_scalar(i, 1)
+
+    return dpnp_cumprod(x1)
+
+
+cpdef dparray dpnp_nancumsum(dparray x1):
+
+    for i in range(x1.size):
```
Comment (samir-nasibli) [inline dpnp/dpnp_algo/dpnp_algo_mathematical.pyx:334]: This is wrong logic, do not modify given array, just check it.

File: dpnp/dpnp_algo/dpnp_algo_mathematical.pyx [hunk 4]
```diff
@@ -344,6 +320,24 @@ cpdef dparray dpnp_multiply(dparray x1, x2):
...
+        if dpnp.isnan(x1[i]):
```
Comment (samir-nasibli) [inline dpnp/dpnp_algo/dpnp_algo_mathematical.pyx:335]: This is wrong logic, do not modify given array, just check it.

File: dpnp/dpnp_algo/dpnp_algo_mathematical.pyx [hunk 5]
```diff
@@ -139,54 +141,28 @@ cpdef dparray dpnp_cross(dparray x1, dparray x2):
     return call_fptr_2in_1out(DPNP_FN_CROSS, x1, x2, x1.shape)
 
 
-cpdef dparray dpnp_cumprod(dparray x1, bint usenan=False):
+cpdef dparray dpnp_cumprod(dparray x1):
+    #instead of x1.shape, (x1.size, ) is passed to the function
```
Comment (samir-nasibli) [inline dpnp/dpnp_algo/dpnp_algo_mathematical.pyx:145]: Please, make this comment section as a docstring, if it is needed or just remove it.

File: dpnp/dpnp_algo/dpnp_algo_mathematical.pyx [hunk 6]
```diff
@@ -139,54 +141,28 @@ cpdef dparray dpnp_cross(dparray x1, dparray x2):
...
-
-    for i in range(result.size):
+    return call_fptr_1in_1out(DPNP_FN_CUMPROD, x1, (x1.size,))
 
-        if not usenan or not dpnp.isnan(x1[i]):
-            cur_res += x1[i]
 
-        result._setitem_scalar(i, cur_res)
+cpdef dparray dpnp_cumsum(dparray x1):
+    #instead of x1.shape, (x1.size, ) is passed to the function
```
Comment (samir-nasibli) [inline dpnp/dpnp_algo/dpnp_algo_mathematical.pyx:157]: Please, make this comment section as a docstring, if it is needed or just remove it.

File: dpnp/dpnp_algo/dpnp_algo_mathematical.pyx [hunk 7]
```diff
@@ -344,6 +320,24 @@ cpdef dparray dpnp_multiply(dparray x1, x2):
         return call_fptr_2in_1out(DPNP_FN_MULTIPLY, x1, x2, x1.shape)
 
 
+cpdef dparray dpnp_nancumprod(dparray x1):
+
+    for i in range(x1.size):
```
Comment (samir-nasibli) [inline dpnp/dpnp_algo/dpnp_algo_mathematical.pyx:325]: The same comment as above, this is wrong logic, do not modify given array, just check it.

File: dpnp/dpnp_iface_mathematical.py [hunk 8]
```diff
@@ -1103,7 +1103,7 @@ def nancumsum(x1, **kwargs):
         if not isinstance(x1, dparray):
```
Comment (samir-nasibli) [inline dpnp/dpnp_iface_mathematical.py:1103]: All you external/third party tests fallbacks to numpy and doesn't check your implementation?
Please write more and more dpnp own tests for this functions.

File: dpnp/backend/kernels/dpnp_krnl_mathematical.cpp [hunk 9]
```diff
@@ -103,6 +103,44 @@ void dpnp_cross_c(void* array1_in, void* array2_in, void* result1, size_t size)
...
+template <typename _DataType_input, typename _DataType_output>
+void dpnp_cumprod_c(void* array1_in, void* result1, size_t size)
+{
+    _DataType_input* array1 = reinterpret_cast<_DataType_input*>(array1_in);
+    _DataType_output* result = reinterpret_cast<_DataType_output*>(result1);
+
+    _DataType_output cur_res = 1;
+
+    for (size_t i = 0; i < size; i++) {
+        cur_res *= array1[i];
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_mathematical.cpp:118]: Do not we need here integer multiply overflow check?

File: dpnp/backend/kernels/dpnp_krnl_mathematical.cpp [hunk 10]
```diff
@@ -103,6 +103,44 @@ void dpnp_cross_c(void* array1_in, void* array2_in, void* result1, size_t size)
...
+template <typename _DataType_input, typename _DataType_output>
+void dpnp_cumsum_c(void* array1_in, void* result1, size_t size)
+{
+    _DataType_input* array1 = reinterpret_cast<_DataType_input*>(array1_in);
+    _DataType_output* result = reinterpret_cast<_DataType_output*>(result1);
+
+    _DataType_output cur_res = 0;
+
+    for (size_t i = 0; i < size; i++) {
+        cur_res += array1[i];
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_mathematical.cpp:137]: The same comment as above, do not we need here overflow check?

File: dpnp/dpnp_algo/dpnp_algo_mathematical.pyx (hunk 5 above)
Comment (densmirn) [reply dpnp/dpnp_algo/dpnp_algo_mathematical.pyx:145]: It looks like a comment for developers (internal comment) for better understanding the code.

Comment (samir-nasibli) [reply dpnp/dpnp_algo/dpnp_algo_mathematical.pyx:145]: This no make sense, this comment includes example. Currently this comments are not so informative so this is needed to rewrite it.

File: dpnp/backend/kernels/dpnp_krnl_mathematical.cpp [hunk 11]
```diff
@@ -103,6 +103,44 @@ void dpnp_cross_c(void* array1_in, void* array2_in, void* result1, size_t size)
     return;
 }
 
+template <typename _KernelNameSpecialization1, typename _KernelNameSpecialization2>
+class dpnp_cumprod_c_kernel;
+
+template <typename _DataType_input, typename _DataType_output>
+void dpnp_cumprod_c(void* array1_in, void* result1, size_t size)
+{
+    _DataType_input* array1 = reinterpret_cast<_DataType_input*>(array1_in);
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_mathematical.cpp:112]:
```suggestion
    const _DataType_input* array1 = reinterpret_cast<const _DataType_input*>(array1_in);
```

Comment (samir-nasibli) [reply dpnp/backend/kernels/dpnp_krnl_mathematical.cpp:112]: Try with `const` please.

File: dpnp/backend/include/dpnp_iface.hpp [hunk 12]
```diff
@@ -169,6 +169,30 @@ INP_DLLEXPORT void dpnp_dot_c(void* array1, void* array2, void* result1, size_t
...
+ * @ingroup BACKEND_API
+ * @brief Custom implementation of cumprod function
+ *
+ * @param [in]  array1_in  Input array.
+ * @param [out] result1    Output array.
+ * @param [in]  size       Number of elements in input arrays.
+ *
+ */
+template <typename _DataType_input, typename _DataType_output>
+INP_DLLEXPORT void dpnp_cumprod_c(void* array1_in, void* result1, size_t size);
```
Comment (samir-nasibli) [inline dpnp/backend/include/dpnp_iface.hpp:182]:
``` suggestion
INP_DLLEXPORT void dpnp_cumprod_c(const void* array1_in, void* result1, const size_t size);
```

File: dpnp/dpnp_algo/dpnp_algo_mathematical.pyx [hunk 13]
```diff
@@ -344,6 +320,28 @@ cpdef dparray dpnp_multiply(dparray x1, x2):
         return call_fptr_2in_1out(DPNP_FN_MULTIPLY, x1, x2, x1.shape)
 
 
+cpdef dparray dpnp_nancumprod(dparray x1):
+
+    cur_x1 = dpnp.copy(x1)
```
Comment (samir-nasibli) [inline dpnp/dpnp_algo/dpnp_algo_mathematical.pyx:325]: You can check in one cycle if given array has `nan` element, if exists then make a copy.

**Review by shssf — APPROVED** [review]

**Comment by shssf** [comment]
I just rebased and squashed this PR to see tests status

**Pull request merged**